nixos/modules/services/autotierfs.nix

```
@@ -0,0 +1,96 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.autotierfs;
+  ini = pkgs.formats.ini { };
+  format = lib.types.attrsOf ini.type;
+  stateDir = "/var/lib/autotier";
+
+  generateConfigName =
+    name: builtins.replaceStrings [ "/" ] [ "-" ] (lib.strings.removePrefix "/" name);
+  configFiles = builtins.mapAttrs (
+    name: val: ini.generate "${generateConfigName name}.conf" val
+  ) cfg.settings;
+
+  getMountDeps =
+    settings: builtins.concatStringsSep " " (builtins.catAttrs "Path" (builtins.attrValues settings));
+
+  mountPaths = builtins.attrNames cfg.settings;
+in
+{
+  options.services.autotierfs = {
+    enable = lib.mkEnableOption "the autotier passthrough tiering filesystem";
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.autotier;
+      defaultText = lib.options.literalExpression "pkgs.autotier";
+      description = "Configured package for the filesystem and its cli.";
+    };
+    settings = lib.mkOption {
+      type = lib.types.nullOr format;
+      default = null;
+      description = ''
+        The contents of the configuration file for autotier.
+        Config can be found in the [autotier repo](https://github.com/45Drives/autotier#configuration)
+      '';
+      example = {
+        "/mnt/autotier" = {
+          Global = {
+            "Log Level" = 1;
+            "Tier Period" = 1000;
+            "Copy Buffer Size" = "1 MiB";
+          };
+          "Tier 1" = {
+            Path = "/mnt/tier1";
+            Quota = "30GiB";
+          };
+          "Tier 2" = {
+            Path = "/mnt/tier2";
+            Quota = "200GiB";
+          };
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.enable -> cfg.settings != null;
```

```suggestion
        assertion = cfg.enable -> cfg.settings != { };
```
I think a better approach would be to apply some nice defaults that can cover this basic configuration, rather than always requiring it. I don't know enough about the service to suggest what though -- or if it's really applicable ofc

---

I prefer to have OOTB defaults in nix services as well, but for autotier the settings really need to be manually configured to specify your different storage tiers (across different mediums/ storage capacities) for running it to make sense in the first place; there's no point in having caching tiers if they're all on the same underlying disk - you'll just get extra latency and performance overhead with no high speed cache benefit.